flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/functions/aggfunctions/RankLikeAggFunctionBase.java

```diff
@@ -0,0 +1,128 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.functions.aggfunctions;
+
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.common.typeinfo.Types;
+import org.apache.flink.table.api.TableException;
+import org.apache.flink.table.expressions.Expression;
+import org.apache.flink.table.expressions.UnresolvedReferenceExpression;
+import org.apache.flink.table.type.DecimalType;
+import org.apache.flink.table.type.InternalType;
+import org.apache.flink.table.type.InternalTypes;
+
+import java.sql.Date;
+import java.sql.Time;
+import java.sql.Timestamp;
+
+import static org.apache.flink.table.expressions.ExpressionBuilder.and;
+import static org.apache.flink.table.expressions.ExpressionBuilder.equalTo;
+import static org.apache.flink.table.expressions.ExpressionBuilder.ifThenElse;
+import static org.apache.flink.table.expressions.ExpressionBuilder.isNull;
+import static org.apache.flink.table.expressions.ExpressionBuilder.literal;
+
+/**
+ * built-in rank like aggregate function, e.g. rank, dense_rank
+ */
+public abstract class RankLikeAggFunctionBase extends DeclarativeAggregateFunction {
+	protected UnresolvedReferenceExpression sequence = new UnresolvedReferenceExpression("sequence");
+	protected UnresolvedReferenceExpression[] lastValues;
+	protected InternalType[] orderKeyTypes;
+
+	public RankLikeAggFunctionBase(InternalType[] orderKeyTypes) {
+		this.orderKeyTypes = orderKeyTypes;
+		lastValues = new UnresolvedReferenceExpression[orderKeyTypes.length];
+		for (int i = 0; i < orderKeyTypes.length; ++i) {
+			lastValues[i] = new UnresolvedReferenceExpression("lastValue_" + i);
+		}
+	}
+
+	@Override
+	public int operandCount() {
+		return orderKeyTypes.length;
+	}
+
+	@Override
+	public TypeInformation getResultType() {
+		return Types.LONG;
+	}
+
+	@Override
+	public Expression[] retractExpressions() {
+		throw new TableException("This function does not support retraction.");
+	}
+
+	@Override
+	public Expression[] mergeExpressions() {
+		throw new TableException("This function does not support merge.");
+	}
+
+	@Override
+	public Expression getValueExpression() {
+		return sequence;
+	}
+
+	protected Expression orderKeyEqualsExpression() {
+		Expression[] orderKeyEquals = new Expression[orderKeyTypes.length];
+		for (int i = 0; i < orderKeyTypes.length; ++i) {
+			Expression lasValue = lastValues[i];
+			// if (lastValue is null) {
```

Comment:
delete these?

Reply:
These are comments, just look like code. I have deleted them.

Reply:
Sorry, my bad, i think it makes sense to keep this as comment, it can help us understand the logic

Reply:
I have reverted them